xdsl/dialects/pdl.py

```diff
@@ -0,0 +1,322 @@
+
+    @staticmethod
+    def get(name: str, args: list[SSAValue]) -> ApplyNativeConstraintOp:
+        return ApplyNativeConstraintOp.build(
+            result_types=[],
+            operands=[args],
+            attributes={"name": StringAttr(name)})
+
+
+@irdl_op_definition
+class ApplyNativeRewriteOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlapply_native_rewrite-mlirpdlapplynativerewriteop
+    """
+    name: str = "pdl.apply_native_rewrite"
+    # https://github.com/xdslproject/xdsl/issues/98
+    # name: OpAttr[StringAttr]
+    args: Annotated[VarOperand, AnyPDLType]
+    res: Annotated[VarOpResult, AnyPDLType]
+
+    def verify_(self) -> None:
+        if 'name' not in self.attributes:
+            raise Exception("ApplyNativeRewriteOp requires a 'name' attribute")
+
+        if not isinstance(self.attributes['name'], StringAttr):
+            raise Exception("expected 'name' attribute to be a StringAttr")
+
+    @staticmethod
+    def get(name: str, args: list[SSAValue],
+            result_types: list[Attribute]) -> ApplyNativeRewriteOp:
+
+        return ApplyNativeRewriteOp.build(
+            result_types=[result_types],
+            operands=[args],
+            attributes={"name": StringAttr(name)})
+
+
+@irdl_op_definition
+class AttributeOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlattribute-mlirpdlattributeop
+    """
+    name: str = "pdl.attribute"
+    value: OptOpAttr[Attribute]
+    value_type: Annotated[OptOperand, TypeType]
+    output: Annotated[OpResult, AttributeType]
+
+
+@irdl_op_definition
+class EraseOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlerase-mlirpdleraseop
+    """
+    name: str = "pdl.erase"
+    op_value: Annotated[Operand, OperationType]
+
+
+@irdl_op_definition
+class OperandOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdloperand-mlirpdloperandop
+    """
+    name: str = "pdl.operand"
+    value_type: Annotated[OptOperand, TypeType]
+    output: Annotated[OpResult, ValueType]
+
+
+@irdl_op_definition
+class OperandsOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdloperands-mlirpdloperandsop
+    """
+    name: str = "pdl.operands"
+    value_type: Annotated[Operand, RangeType[TypeType]]
+    output: Annotated[OpResult, RangeType[ValueType]]
+
+
+@irdl_op_definition
+class OperationOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdloperation-mlirpdloperationop
+    """
+    name: str = "pdl.operation"
+    opName: OptOpAttr[StringAttr]
+    attributeValueNames: OpAttr[ArrayAttr[StringAttr]]
+
+    operandValues: Annotated[VarOperand, ValueType | RangeType[ValueType]]
+    # in PDL docs, this is just a handle to AttributeType, not a range.
+    # Why is it different to operandvalues
+    attributeValues: Annotated[VarOperand,
+                               AttributeType | RangeType[AttributeType]]
```

Comment:
While I still did not quite understand why they designed it that way, I don't think we want to leave this message here